### PR TITLE
feat: Add user configuration for all providers

### PR DESCRIPTION
**Pull request by dlipovetsky**
- Defines a cluster-level variable for defining one or more users
- Patches bootstrap templates for control plane and worker node pools with user configuration
- Derives value of CABPK's `User.LockPassword`

### CODE REVIEW

File: api/v1alpha1/clusterconfig_types.go [hunk 1]
```diff
@@ -344,6 +348,66 @@ func (ImageRegistries) VariableSchema() clusterv1.VariableSchema {
...
 }
 
+type Users []User
+
+func (Users) VariableSchema() clusterv1.VariableSchema {
+	return clusterv1.VariableSchema{
+		OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+			Description: "Users to add to the machine",
+			Type:        "array",
+			Items:       ptr.To(User{}.VariableSchema().OpenAPIV3Schema),
```
Comment (faiq) [inline api/v1alpha1/clusterconfig_types.go:358]: Why ptr.To?

File: api/v1alpha1/clusterconfig_types.go [hunk 2]
```diff
@@ -344,6 +348,66 @@ func (ImageRegistries) VariableSchema() clusterv1.VariableSchema {
...
+}
+
+// User defines the input for a generated user in cloud-init.
+type User struct {
+	// Name specifies the user name
+	Name string `json:"name"`
+
+	// Passwd specifies a hashed password for the user
+	// +optional
+	Passwd *string `json:"passwd,omitempty"`
```
Comment (faiq) [inline api/v1alpha1/clusterconfig_types.go:370]: for some optional fields we use just a string, and for this we're using a pointer. 

i forgot the reasoning behind not using a pointer, but this seems to be a convention in most projects. i'd like to get some consensus and land on what we're doing.

File: pkg/handlers/generic/mutation/users/inject.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+	variableName string,
+	variableFieldPath ...string,
+) *usersPatchHandler {
+	return &usersPatchHandler{
+		variableName:      variableName,
+		variableFieldPath: variableFieldPath,
+	}
+}
+
+func (h *usersPatchHandler) Mutate(
```
Comment (faiq) [inline pkg/handlers/generic/mutation/users/inject.go:51]: i like how this function handles users generally. nicely done.

File: api/v1alpha1/clusterconfig_types.go (hunk 2 above)
Comment (dlipovetsky) [reply api/v1alpha1/clusterconfig_types.go:370]: I think this a good explanation:

> Using a pointer allows distinguishing unset from the zero value for that type. There are some cases where, in principle, a pointer is not needed for an optional field since the zero value is forbidden, and thus implies unset. There are examples of this in the codebase. However:
> 
> it can be difficult for implementors to anticipate all cases where an empty value might need to be distinguished from a zero value
> structs are not omitted from encoder output even where omitempty is specified, which is messy;
> having a pointer consistently imply optional is clearer for users of the Go language client, and any other clients that use corresponding types
> 
> Therefore, we ask that pointers always be used with optional fields that do not have a built-in nil value.
> -- https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required explains the need for using a pointer

Comment (faiq) [reply api/v1alpha1/clusterconfig_types.go:370]: I see Passwd can be `""` which is not the same as having no password.

File: api/v1alpha1/clusterconfig_types.go (hunk 1 above)
Comment (dlipovetsky) [reply api/v1alpha1/clusterconfig_types.go:358]: Good question. We can't just reference the return value of `VariableSchema()`, although this has been [proposed](https://github.com/golang/go/issues/45653), but apparently turned down.

File: api/v1alpha1/clusterconfig_types.go (hunk 2 above)
Comment (dlipovetsky) [reply api/v1alpha1/clusterconfig_types.go:370]: That's what I thought at first... but it turns out that an empty string is invalid here, because it's eventually passed to cloud-init's `passwd` field, which requires the _hash_ of a password. And the hash can never be an empty string.

If we make this a string, rather than a pointer to a string, then the zero value (empty string) won't be marshalled, which is actually what we want.

Comment (jimmidyson) [inline api/v1alpha1/clusterconfig_types.go:370]: Is this is a hashed password, can we call it `hashedPassword` or `passwordHash`?
```suggestion
	HashedPasswordd *string `json:"hashedPassword,omitempty"`
```

Comment (jimmidyson) [reply api/v1alpha1/clusterconfig_types.go:370]: I'm not clear from the discussion above whether we want to keep this as a pointer or not? I don't think we need this to be a pointer as it has a valid zero value that will not be marshalled.

File: api/v1alpha1/clusterconfig_types.go [hunk 4]
```diff
@@ -344,6 +348,66 @@ func (ImageRegistries) VariableSchema() clusterv1.VariableSchema {
...
+func (User) VariableSchema() clusterv1.VariableSchema {
+	return clusterv1.VariableSchema{
+		OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+			Type: "object",
+			Properties: map[string]clusterv1.JSONSchemaProps{
+				"name": {
+					Description: "The username",
+					Type:        "string",
+				},
+				"passwd": {
```
Comment (jimmidyson) [inline api/v1alpha1/clusterconfig_types.go:390]: I'd prefer to be explicit that this is a hashed password
```suggestion
				"hashedPasswordd": {
```

File: pkg/handlers/generic/mutation/users/doc.go [hunk 5]
```diff
@@ -0,0 +1,5 @@
+// Copyright 2023 D2iQ, Inc. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters,verbs=watch;list;get
```
Comment (jimmidyson) [inline pkg/handlers/generic/mutation/users/doc.go:4]: Why is this needed? I don't think handler performs any k8s API calls?

File: pkg/handlers/generic/mutation/users/inject.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+	// LockPassword is not part of our API, because we can derive its value
+	// for the use cases our API supports.
+	//
+	// We do not support the edge cases where a password is defined, but
+	// password authentication is disabled, or where no password is defined, but
+	// password authentication is enabled.
+	//
+	// We disable password authentication by default.
+	bootstrapUser.LockPassword = ptr.To[bool](true)
```
Comment (jimmidyson) [inline pkg/handlers/generic/mutation/users/inject.go:135]: Type should be inferrable
```suggestion
	bootstrapUser.LockPassword = ptr.To(true)
```

File: pkg/handlers/generic/mutation/users/inject.go [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+	if userFromVariable.Passwd != nil {
+		// We enable password authentication only if a password is defined.
+		bootstrapUser.LockPassword = ptr.To[bool](true)
```
Comment (jimmidyson) [inline pkg/handlers/generic/mutation/users/inject.go:138]: Type should be inferrable
```suggestion
		bootstrapUser.LockPassword = ptr.To(true)
```

File: pkg/handlers/generic/mutation/users/tests/generate_patches.go [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+	"github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/api/v1alpha1"
+	"github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/common/pkg/capi/clustertopology/handlers/mutation"
+	"github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/common/pkg/testutils/capitest"
+	"github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/common/pkg/testutils/capitest/request"
+)
+
+var (
+	testUser1 = v1alpha1.User{
+		Name:   "complete",
+		Passwd: ptr.To[string]("password"),
```
Comment (jimmidyson) [inline pkg/handlers/generic/mutation/users/tests/generate_patches.go:23]: Type should be inferrable
```suggestion
		Passwd: ptr.To("password"),
```

File: pkg/handlers/generic/mutation/users/tests/generate_patches.go [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+		SSHAuthorizedKeys: []string{
+			"key1",
+			"key2",
+		},
+		Sudo: ptr.To[string]("ALL=(ALL) NOPASSWD:ALL"),
```
Comment (jimmidyson) [inline pkg/handlers/generic/mutation/users/tests/generate_patches.go:28]: Type should be inferrable
```suggestion
		Sudo: ptr.To("ALL=(ALL) NOPASSWD:ALL"),
```

File: pkg/handlers/generic/mutation/users/variables_test.go [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+		false,
+		clusterconfig.NewVariable,
+		// HTTPProxy
+		capitest.VariableTestDef{
+			Name: "valid users",
+			Vals: v1alpha1.GenericClusterConfig{
+				Users: []v1alpha1.User{
+					{
+						Name:   "complete",
+						Passwd: ptr.To[string]("password"),
```
Comment (jimmidyson) [inline pkg/handlers/generic/mutation/users/variables_test.go:30]: Type should be inferrable
```suggestion
						Passwd: ptr.To("password"),
```

File: pkg/handlers/generic/mutation/users/variables_test.go [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+						SSHAuthorizedKeys: []string{
+							"key1",
+							"key2",
+						},
+						Sudo: ptr.To[string]("ALL=(ALL) NOPASSWD:ALL"),
```
Comment (jimmidyson) [inline pkg/handlers/generic/mutation/users/variables_test.go:35]: Type should be inferrable
```suggestion
						Sudo: ptr.To("ALL=(ALL) NOPASSWD:ALL"),
```

**Review by jimmidyson — COMMENTED** [review]
Few nits but looking good!

File: api/v1alpha1/clusterconfig_types.go (hunk 2 above)
Comment (dlipovetsky) [reply api/v1alpha1/clusterconfig_types.go:370]: I started by using the field names in the CABPK types, but I agree, `HashedPassword` is an improvement.

Comment (dlipovetsky) [reply api/v1alpha1/clusterconfig_types.go:370]: > I don't think we need this to be a pointer as it has a valid zero value that will not be marshalled.

What I tried to say abovei s that the zero value is _not_ valid: An empty string cannot be a hashed password.

I'll change it from a pointer to a value.

CABPK supports both cloud-init and ignition for bootstrap, and both take a hashed password as input:
- https://github.com/kubernetes-sigs/cluster-api/blob/0624ec0790d1190fcb980c4816cd38aa430e4b87/bootstrap/kubeadm/internal/cloudinit/users.go#L24-L26
- https://github.com/kubernetes-sigs/cluster-api/blob/7bb4cde098319db9378a4f327294c96d645e892f/bootstrap/kubeadm/internal/ignition/clc/clc.go#L76-L78

**Review by dkoshkin — DISMISSED** [review]
LGTM

File: api/v1alpha1/clusterconfig_types.go [hunk 12]
```diff
@@ -344,6 +348,66 @@ func (ImageRegistries) VariableSchema() clusterv1.VariableSchema {
...
+					Description: "The hashed password for the user",
+					Type:        "string",
+				},
+				"sshAuthorizedKeys": {
```
[the remark this answers is not in the document]
Comment (dkoshkin) [reply api/v1alpha1/clusterconfig_types.go:395]: Better yet, a doc in https://github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/tree/main/docs/content/customization/generic would be great!

[the remark this answers is not in the document]
Comment (dlipovetsky) [reply api/v1alpha1/clusterconfig_types.go:395]: Sure, I'll expand the description.

[the remark this answers is not in the document]
Comment (dlipovetsky) [reply api/v1alpha1/clusterconfig_types.go:399]: Done. LMK what you think

File: api/v1alpha1/clusterconfig_types.go (hunk 2 above)
Comment (dlipovetsky) [reply api/v1alpha1/clusterconfig_types.go]: Changed to value.

Comment (jimmidyson) [reply api/v1alpha1/clusterconfig_types.go]: Perfect thank you!

File: api/v1alpha1/clusterconfig_types.go [hunk 13]
```diff
@@ -344,6 +348,71 @@ func (ImageRegistries) VariableSchema() clusterv1.VariableSchema {
...
+					Description: "A list of SSH authorized keys for this user",
+					Type:        "array",
+					Items: &clusterv1.JSONSchemaProps{
+						// No description, because the one for the parent array is enough.
+						Type: "string",
+					},
+				},
+				"sudo": {
+					Description: "The sudo rule that applies to this user",
+					Type:        "string",
```
Comment (jimmidyson) [inline api/v1alpha1/clusterconfig_types.go:409]: I wonder if there is a format/pattern that could be applied here to prevent invalid config? Could be done as a follow-up or not at all...

File: api/v1alpha1/clusterconfig_types.go [hunk 14]
```diff
@@ -344,6 +348,71 @@ func (ImageRegistries) VariableSchema() clusterv1.VariableSchema {
...
+}
+
+func (User) VariableSchema() clusterv1.VariableSchema {
+	return clusterv1.VariableSchema{
+		OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+			Type: "object",
+			Properties: map[string]clusterv1.JSONSchemaProps{
+				"name": {
+					Description: "The username",
+					Type:        "string",
```
Comment (jimmidyson) [inline api/v1alpha1/clusterconfig_types.go:393]: Name is required and should not accept empty strings. I wonder if we can specify a default here?

File: api/v1alpha1/clusterconfig_types.go [hunk 15]
```diff
@@ -344,6 +348,71 @@ func (ImageRegistries) VariableSchema() clusterv1.VariableSchema {
...
+
+	// SSHAuthorizedKeys specifies a list of public SSH keys to write to the
+	// machine. Use the corresponding private SSH keys to authenticate. See SSH
+	// documentation for instructions to create a key pair.
+	// +optional
+	SSHAuthorizedKeys []string `json:"sshAuthorizedKeys,omitempty"`
+
+	// Sudo specifies a sudo role for the user
+	// +optional
+	Sudo *string `json:"sudo,omitempty"`
```
Comment (jimmidyson) [inline api/v1alpha1/clusterconfig_types.go:383]: Doesn't the same argument re pointer/value apply for this as `HashedPassword`?

File: api/v1alpha1/clusterconfig_types.go [hunk 16]
```diff
@@ -344,6 +348,71 @@ func (ImageRegistries) VariableSchema() clusterv1.VariableSchema {
...
+}
+
+func (User) VariableSchema() clusterv1.VariableSchema {
+	return clusterv1.VariableSchema{
+		OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+			Type: "object",
+			Properties: map[string]clusterv1.JSONSchemaProps{
```
Comment (jimmidyson) [inline api/v1alpha1/clusterconfig_types.go:390]: Let's specify at least name as required:

```suggestion
			Required: []string{"name"},
			Properties: map[string]clusterv1.JSONSchemaProps{
```

File: pkg/handlers/generic/mutation/users/inject.go [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+	// We do not support these edge cases:
+	// (a) Hashed password is defined, password authentication is not enabled.
+	// (b) Hashed password is not defined, password authentication is enabled.
+	//
+	// We disable password authentication by default.
+	bootstrapUser.LockPassword = ptr.To(true)
+	if userFromVariable.HashedPassword != "" {
+		// We enable password authentication only if a hashed password is defined.
+		bootstrapUser.LockPassword = ptr.To(true)
+	}
```
Comment (jimmidyson) [inline pkg/handlers/generic/mutation/users/inject.go:139]: This is currently always setting `LockPassword` to true, whether `HashedPassword` is set or not. This can be fixed and simplified to:

```suggestion
	bootstrapUser.LockPassword = ptr.To(userFromVariable.HashedPassword == "")
```

File: api/v1alpha1/clusterconfig_types.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (jimmidyson) [reply api/v1alpha1/clusterconfig_types.go:399]: Could you add this to a doc in `docs/content/customization/generic` please?

File: api/v1alpha1/clusterconfig_types.go (hunk 14 above)
Comment (dlipovetsky) [reply api/v1alpha1/clusterconfig_types.go:393]: I don't think a default makes sense. Even if we assume there is always one entry for the "default" user in a machine image, the actual name is not universal.

Comment (jimmidyson) [reply api/v1alpha1/clusterconfig_types.go:393]: Still please make it required.

File: pkg/handlers/generic/mutation/users/inject.go (hunk 17 above)
Comment (jimmidyson) [reply pkg/handlers/generic/mutation/users/inject.go:139]: Actually I think this indicates a missing test?

File: api/v1alpha1/clusterconfig_types.go (hunk 13 above)
Comment (dlipovetsky) [reply api/v1alpha1/clusterconfig_types.go:409]: Valid config is definitely desirable, because sudo fails if any rule has invalid syntax, preventing anyone (not just the user that introduced the invalid configuration) from using sudo.

Sudo rules are defined using an EBNF grammar. We would need a parser to validate a rule.

Comment (jimmidyson) [reply api/v1alpha1/clusterconfig_types.go:409]: That seems overkill 😅 so probably not worth it

File: api/v1alpha1/clusterconfig_types.go [hunk 18]
```diff
@@ -344,6 +348,71 @@ func (ImageRegistries) VariableSchema() clusterv1.VariableSchema {
...
+				"name": {
+					Description: "The username",
+					Type:        "string",
+				},
+				"hashedPassword": {
+					Description: "The hashed password for the user",
+					Type:        "string",
```
[the remark this answers is not in the document]
Comment (dlipovetsky) [reply api/v1alpha1/clusterconfig_types.go:397]: Good point. Fortunately, there's a well-defined regex in the `crypt` [man page](https://man.archlinux.org/man/crypt.5#Hashed)! I'll add it.

[the remark this answers is not in the document]
Comment (jimmidyson) [reply api/v1alpha1/clusterconfig_types.go:397]: oh wow! neat!

File: pkg/handlers/generic/mutation/users/inject.go (hunk 17 above)
Comment (dlipovetsky) [reply pkg/handlers/generic/mutation/users/inject.go:139]: Thanks. I fixed this and added a unit test.

File: api/v1alpha1/clusterconfig_types.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (dlipovetsky) [reply api/v1alpha1/clusterconfig_types.go:397]: I spoke too soon. There are _many_ well-defined regexes, one for each supported password hash function.

The set of supported functions depends on the distro, i.e. the machine image used.

To validate this, we'd need to validate against a list of known hash functions. If the user chooses a function we don't support, but that their machine does, we'll give a false negative.

I don't think we can do validation here.

[the remark this answers is not in the document]
Comment (dlipovetsky) [reply api/v1alpha1/clusterconfig_types.go:399]: I've added a comment to explain why we don't validate the input.

File: api/v1alpha1/clusterconfig_types.go (hunk 13 above)
Comment (dlipovetsky) [reply api/v1alpha1/clusterconfig_types.go:418]: Added a comment explaining why we don't validate the input

File: api/v1alpha1/clusterconfig_types.go [hunk 19]
```diff
@@ -395,8 +395,15 @@ func (User) VariableSchema() clusterv1.VariableSchema {
...
-					Description: "The hashed password for the user",
+					Description: "The hashed password for the user. Must be in the format of some hash function supported by the OS.",
 					Type:        "string",
+					// Although the crypt (5) man page lists regexes for
+					// supported hash functions. Although we could validate
+					// input against a set of regexes, this set may be different
+					// from the set supported by the chosen OS, and we could
+					// return a false negative or positive. For this reason, we
+					// do not validate the input.
+					//
```
Comment (dkoshkin) [inline api/v1alpha1/clusterconfig_types.go:406]: Were the  2 "Although" intentional?
Also "//" at the end

File: api/v1alpha1/clusterconfig_types.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (jimmidyson) [reply api/v1alpha1/clusterconfig_types.go:399]: Ok thanks

**Comment by dlipovetsky** [comment]
Screenshot_28-3-2024_154051_localhost

Screenshot of the new doc :smile:

File: api/v1alpha1/clusterconfig_types.go (hunk 19 above)
Comment (dlipovetsky) [reply api/v1alpha1/clusterconfig_types.go]: Nope, typo. Fixed it. Thanks!

File: api/v1alpha1/clusterconfig_types.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (dlipovetsky) [reply api/v1alpha1/clusterconfig_types.go:408]: Done.

Here's a semi-readable screenshot:

image

**Comment by dlipovetsky** [comment]
I changed the fixup commit messages to make them conform to the precommit linter, then force-pushed.

**Comment by dlipovetsky** [comment]
precommit fails because markdownlint complains about line length >= 120 chars in the documentation I wrote.

While maximum line length make sense for a programming language, I think it makes less sense for prose:
1. The docs are written in a text editor, and almost any editor can automatically wrap lines.  I don't want to manually wrap lines the first time, or re-wrap lines as I update prose.
1. The docs are read in a browser and a text editor. Almost any browser or editor can automatically wrap lines. Manual wrapping can make it more difficult to wrap text automatically for a given viewport size.

I'd like to [remove the rule](https://github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/blob/0857199694980fbaf3325d365389b4fb8a1aafed/.markdownlint.json#L4), and eventually revert all docs to automatically wrapped lines.

**Comment by dkoshkin** [comment]
>I'd like to remove the rule, and eventually revert all docs to automatically wrapped lines.

Thanks, I support this.

**Review by dkoshkin — APPROVED** [review]

**Comment by dlipovetsky** [comment]
> > I'd like to remove the rule, and eventually revert all docs to automatically wrapped lines.
> 
> Thanks, I support this.

Thanks, I filed https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/issues/480.

**Review by faiq — APPROVED** [review]

**Review by supershal — APPROVED** [review]
